Add new module with http4s-grpc implementation

sgjbryan opened the pull request:
Enables use of gRPC without pulling in all the Java dependencies

- Proto definitions taken from the [OLTP spec](https://github.com/open-telemetry/opentelemetry-proto/tree/v1.0.0/opentelemetry/proto)
- Transformation from trace4cats models to OLTP protos largely taken from the existing JSON-based OLTP/HTTP implementation

janstenpickle commented:
Thanks @sgjbryan. Rather than including the proto directly, do you know if it's maven? If so, we can use ScalaPB [to add a dependency](https://scalapb.github.io/docs/third-party-protos)

sgjbryan commented:
Ah good flag. It looks like there is a [maven artifact](https://mvnrepository.com/artifact/io.opentelemetry.proto/opentelemetry-proto) but only with `alpha` versions. May still be better to use `1.0.0-alpha` rather than manually including the proto files here, what do you think?

janstenpickle commented:
Hmm I think we should probably avoid using the alpha versions, it's weird that it's only those published in maven though, maybe an unpstream release issue?

Anyway, I'm wondering if we should do a submodule for these so we can pin that to a particular version, assuming we can make that with the build here of course. If that ends up being too much effort, what you've done here is fine.

janstenpickle commented on modules/opentelemetry-otlp-http4s-grpc-exporter/src/main/scala/trace4cats/opentelemetry/otlp/proto/ExportSpansRequest.scala:
```diff
@@ -0,0 +1,87 @@
+        .view
+        .map { case (service, spans) =>
+          ResourceSpans(
+            resource = Resource(attributes =
+              List(KeyValue("service.name", AnyValue(AnyValue.Value.StringValue(service)).some))
+            ).some,
```
See the recently merged #260, this should probably include all of the `TraceProcess` attributes.

janstenpickle commented on modules/opentelemetry-otlp-http4s-grpc-exporter/src/main/scala/trace4cats/opentelemetry/otlp/OpenTelemetryOtlpHttp4sGrpcSpanExporter.scala:
```diff
@@ -0,0 +1,35 @@
+import trace4cats.model.Batch
+import trace4cats.opentelemetry.otlp.proto.ExportSpansRequest
+
+object OpenTelemetryOtlpHttp4sGrpcSpanExporter {
```
Could you also mirror the host/port constructor like [here](https://github.com/trace4cats/trace4cats-opentelemetry/blob/master/modules/opentelemetry-otlp-grpc-exporter/src/main/scala/trace4cats/opentelemetry/otlp/OpenTelemetryOtlpGrpcSpanExporter.scala#L12-L17) please?

janstenpickle commented on modules/opentelemetry-otlp-http4s-grpc-exporter/src/main/scala/trace4cats/opentelemetry/otlp/OpenTelemetryOtlpHttp4sGrpcSpanExporter.scala:
```diff
@@ -0,0 +1,35 @@
+    uri: Uri,
+    staticHeaders: List[Header.ToRaw] = List.empty
+  ): SpanExporter[F, G] = new SpanExporter[F, G] {
+    val service = TraceService.fromClient[F](client, uri)
```
Not that it matters massively, but this should probably be `private`

janstenpickle reviewed (commented):
Looks good, just a couple of nitpicks!

sgjbryan commented:
> Hmm I think we should probably avoid using the alpha versions, it's weird that it's only those published in maven though, maybe an unpstream release issue?

Yeah, it seems like a release issue somewhere - a lot of the otel artifacts seem to only be published with the `-alpha` suffix.

> Anyway, I'm wondering if we should do a submodule for these so we can pin that to a particular version, assuming we can make that with the build here of course. If that ends up being too much effort, what you've done here is fine.

Not sure I quite follow what you mean here. As in just a separate SBT submodule here which contains _only_ the proto files, which is then depended on by this new exporter module?

janstenpickle commented:
> Not sure I quite follow what you mean here. As in just a separate SBT submodule here with a fixed version which contains only the proto files, which is then depended on by this new exporter module?

Oh I mean a git submodule to the otel protos, that way we don't have to copy/paste them into this repo

sgjbryan commented:
Ah gotcha, that makes sense. All _should_ be working now:
- Using a git submodule for the proto dependencies
- Updated the GitHub workflows to checkout submodules (think this will be necessary for the build to work)
- Addressed other PR comments

janstenpickle commented:
Thanks @sgjbryan, looks like the compiler fatal warnings may need to be turned off for the generate proto code 😢

sgjbryan commented:
Ah, looks like `http4s-grpc` is [only cross-published for 2.13 and 3.2](https://github.com/davenverse/http4s-grpc/blob/1b18484c52978d3703c2293b4b2eaeff26035b0c/build.sbt#L17)... Let me see what I can do

janstenpickle commented on build.sbt:
```diff
@@ -86,3 +88,36 @@ lazy val `opentelemetry-otlp-http-exporter` =
+      name := "trace4cats-opentelemetry-otlp-http4s-grpc-exporter",
+      libraryDependencies ++= Seq(Dependencies.trace4catsCore),
+      Compile / PB.targets ++= Seq(scalapb.gen(grpc = false) -> (Compile / sourceManaged).value / "scalapb"),
+      scalacOptions += {
+        CrossVersion.partialVersion(scalaVersion.value) match {
+          case Some((2, _)) => "-Wconf:src=proto/.*:silent"
+          case _ => ""
+        }
+      },
+      scalacOptions -= {
+        CrossVersion.partialVersion(scalaVersion.value) match {
+          case Some((2, _)) => ""
+          case _ => "-Wvalue-discard"
+        }
+      }
```
Gotta love Scala 3...

janstenpickle reviewed (approved):
LGTM, thanks!
